### ISSUE TITLE
Consuming an unmanaged Scala 3.1.1 JAR fails in a 2.13.8 SBT project

### ISSUE THREAD

**Issue by hilcode** [sbt/sbt#6865]
## steps

Run `sbt compile` and observe the `typer` error. If you change the `scalaVersion` in `library/build.sc` everything works fine.

## problem

When consuming a Scala 3 compiled library _as an unmanaged JAR_, TASTY doesn't seem to work correctly or is not called correctly.

## expectation

This should work for both `2.13.8` and `3.1.1`.

## notes
[sbt-bug.zip](https://github.com/sbt/sbt/files/8433571/sbt-bug.zip)

Unzip `sbt-bug.zip` and create `lib/library.jar` as a symlink to `library/out/library/jar.dest/out.jar`. So (while in `lib`) `ln -sf ../library/out/library/jar.dest/out.jar library.jar`.

**Comment by SethTisue** [comment]
might interest @bishabosha?

**Comment by bishabosha** [comment]
so I ran `sbt compile` and I am getting an error in the `TreeUnpickler` code, (implying we are reading TASTy files) - so I will check further - this could be a scala compiler bug rather than related to sbt.

Edit: update - so when I turn library into a standard project depended upon from with `dependsOn` (still compiled by 3.1.1), then it works :/

```
sbt:SBT Bug> root/dependencyTree
[info] sbt-bug:sbt-bug_2.13:0.1.0-SNAPSHOT [S]
[info]   +-library:library_3:0.1.0-SNAPSHOT
[info]     +-org.scala-lang:scala3-library_3:3.1.1 [S]
```

the tasty files also seem identical.

Edit2: I ran with `-Ydebug-tasty -Ydebug` and the case with the unmanaged jar starts trying to complete `module val Library` at `Addr(4)` before it's finished reading the top level symbols - Possibly reading the annotations is forcing some completion but this would be quite strange to me, unless packages are forced differently depending on if they come from jar vs directory. I can't tell for sure until I recreate this with a locally built compiler

**Comment by hilcode** [comment]
@bishabosha Indeed, when using `publishM2Local` in Mill and depending on that JAR "normally", everything works fine. The JAR itself is byte for byte identical, including the `.tasty` file (AFAICT).

**Comment by eed3si9n** [comment]
I am closing this due to inactivity, but if this is still relevant let us know.
